app/controllers/authentication.js

```diff
@@ -115,105 +117,166 @@ var Controller = function(){
+              message: 'Did not get email address.',
+            });
+          }
+
+          callback(null, accessToken, headers, profile);
+        });
+    };
+
+    var checkOrganization = function (accessToken, headers, profile, callback) {
+      request.get({ url: userApiUrl + '/orgs', qs: accessToken, headers: headers, json: true },
+        function (err, response) {
+          if (err) { callback(err); return; }
+
+          var belongsOrg = _.find(response.body, { login: nconf.get('github').organization });
+
+          if (belongsOrg) {
+            callback(null, accessToken, headers, profile);
+          } else {
+            return res.status(401).send({
+              message: 'You do not have required access.',
+            });
+          }
+        });
+    };
+
+    var checkAdmin = function (accessToken, headers, profile, callback) {
+      request.get({ url: userApiUrl + '/teams', qs: accessToken, headers: headers, json: true },
+        function (err, response) {
+          if (err) { callback(err); return; }
+
+          var isAdmin = _.find(response.body, function (team) {
+            return team.name === nconf.get('github').adminTeam && team.organization.login === nconf.get('github').organization;
           });
-        }
-        if( !profile.email ){
-          console.log(response.headers);
-          console.log(profile);
-          return res.status(409).json({
-            message: 'did not get email address'
-          }); 
-        }
-        // Step 3a. Link user accounts.
-        if (req.headers.authorization ) {
-          console.log('Link user account')
-          console.log(profile)
-          User.findOne({ github: profile.id }, function(err, existingUser) {
-            if (existingUser) {
-              console.log(existingUser)
-              return res.status(409).send({ message: 'There is already a GitHub account that belongs to you' });
-            }
-            var token = req.headers.authorization.split(' ')[1];
-            var payload = jwt.decode(token, config.TOKEN_SECRET);
-            User.findById(payload.sub, function(err, user) {
-              if (!user) {
-                return res.status(400).send({ message: 'User not found' });
-              }
-              user.github = profile.id;
-              user.picture = user.picture || profile.avatar_url;
-              user.displayName = user.displayName || profile.name;
-              user.name = user.displayName;
-              user.email = user.email;
-              user.save(function(error) {
-                var token = createJWT(user);
-                res.send({ token: token });
-              });
+          console.log(isAdmin);
+          profile.role = isAdmin ? 'admin' : 'user';
+
+          callback(null, profile);
+        });
+    };
+
+    var ready = function (err, profile) {
+      // Helper function to update user's admin group status.
+      var updateGroup = function (user, done) {
```

Comment:
maybe these should be inside another waterfall, or even inside User&/Group Model..